### DISCUSSION TITLE
onHover for member of CompositeLayer is not called?

### DISCUSSION THREAD

**Discussion by cneumann** [visgl/deck.gl#5497] [Q&A] [answered]
I'm writing a CompositeLayer that returns a single LineLayer from its `renderLayers` function. I would like my CompositeLayer to encapsulate the hover behavior and therefore set the `onHover` prop of the LineLayer:

```javascript
function handleOnHover(info, event) {
  // ...

  // indicate that hovering is handled and should not bubble up to Deck
  return true;
}

class MyLayer extends CompositeLayer {
  renderLayers() {
    return [
      new LineLayer(this.getSubLayerProps({
        id: 'line',
        data: //...
        pickable: true,
        onHover: handleOnHover
      });
    ];
  }
}
```
Unfortunately `handleOnHover` is never called. Is this the intended behavior with CompositeLayer?

If I instead pass a `onHover` prop to the MyLayer instance I pass to the DeckGL component, that callback is invoked, but that means the user of MyLayer has to deal with setting up hovering functionality and I'd rather have it encapsulated in the layer. 

What I want to do with this: I want to highlight a subset of the lines based on which line is being hovered - in my data there are certain lines that are related and I want to highlight all related ones when hovering over one of them. I have the necessary data structures to determine these related lines and it feels like this should be something that is an implementation detail of `MyLayer`.

**Comment by Pessimistress** [comment, marked as answer]
Yes, this is intentional. The sub layers are considered internal to a composite layer's implementation and are by default not exposed to the end user. For example, if you create a TileLayer that renders a GeoJsonLayer that renders scatterplot and polygon layers, when you specify `onHover` on the TileLayer, it is called only once per mouse move, instead of 3 times bubbling up from the lowest level.

When you implement your own layer class, a picking event is handled via the `getPickingInfo` method, as explained here: https://deck.gl/docs/developer-guide/custom-layers/picking#event-propagation This method on the CompositeLayer receives an additional `sourceLayer` argument that indicates where the event comes from. You may add any additional logic here, including invoking custom callbacks if you need to.

**Reply by cneumann** [reply]
Thanks for the quick reply. I'll take a look at `getPickingInfo`. In the meantime I found a workaround by overriding the `onHover` member function of Layer in `MyLayer`.
